Pan/Zoom and Variable X Axis; many other changes

mackworth opened the pull request:
### Summary
Pan/Zoom and Variable X Axis; many other changes

### Fixes Issues
This pull request fixes the following issues:  
 - #296, #286, #282, #278, #271, #269, #261 
 - Handles very large numbers of points (realistically up to thousands) 
 - Restore API back to NSInteger (for Swift) and resolved internal changes 
 - Fix test for white/alpha to handle extended Grey space 
 - Animation without "display dots during" left dots off previously 
 - Remove prev reference lines if axis turned off 
 - Supports bezier curves even when line isn't being drawn 
 - Top/Bottom reference lines were outside box 
 - Fix bug with `touchLineInputColor` (doesn't change after initial setting) 
 - `averageLine` color picked up from line if not set  
 - NoDataLabel color should not default from Line (which defaults to white, same as background) 
 - If label isn't used after initially being created, ensure it's removed from view 
 - Avoid possible infinite loop if delegate gives a zero incrementIndex for x-axis 
 - Fix one-pixel gap between y-axis and graph 
 - Remove spurious NSLogs 
 -  If x-axis background is defaulting to `colorBottom`, then also use `alphaBottom` to match; same for y-axis and `colorTop`. 
 - If neither X nor Y reference lines, set line's `enableRefLines` to `NO` (although default, might have been previously `YES`) 
 - One point is handled by main routines rather than special case 
 - Simplified handling of zero points 
 - Many `BEMNull` related bugs: 
   - If null data and interpolation on, then extrapolate for beginning/ending nulls. 
   - Ensure all BEMNulls do not crash 
   - When Bezier off and Interpolate nulls off, properly draws line segments now 
   - If interpolation off, then bezier line should be interrupted by gaps (but not top/bottom). 
   - If data changes to null value, ensure corresponding Dot isn't left on chart

### Changes
The following changes are included in this pull request:  
 - Pan/Zoom
 - Variable X Axis
 - Cubic Bezier Algorithm
 - Many others (see below for details)
 - New Features:
   - Optional user-scaling pan/zoom with double-tap to restore
   - Variable x-axis
     - Adds new delegates: 
       - `lineGraph:locationForPointAtIndex:` 
       - `lineGraph:labelOnXAxisForLocation:` 
       - `numberOfXAxisLabelsOnLineGraph:maxXValueForLineGraph:` 
       - `minXValueForLineGraph`
   - Ability to take screen shot bigger (or smaller) than screen
   - New delegate method to specify y-axis labels
   - Axis at top option
   - Cubic Bezier algorithm for smoother lines
   - Implemented numberOfXAxisLabelsOnLineGraph for indexed xAxis 
   - TestBed usable on iPhone and Split View
   - NumberOfPoints entry in TestBed
   - Activity display in TestBed to test didFinish
   - Upward-compatible API

### Notes
- [x] This pull request makes breaking changes, and if so, involves the following:  
  - [ ] Public API availability  
  - [x] Internal functionality or behavior  
    - All global vars changed to internal properties  
    - All MutableArrays to Arrays  
    - Use Views to group dots and labels over BEMLine.  
    - Set size of major views at beginning, so all calcs based on those views  
    - Moves getting data and calculations of points to new getData method  
- [x] I have run and ensured that this pull request passes all XCTests

mackworth committed:
Add encoder/decoder for BEMLineGraph properties
Add averageValue delegate method to allow BEMLineGraph to be separate from BEMCalculator
Restore modifyPopUpView delegate method
Use caption preferredFont instead of SystemFont fro NoData
Handle overlapping labels better
Reuse code for popup and permanent labels

Bug fixes
	ensure chart doesn't grow views when zero or one data points

mackworth committed:
Add Types to all Arrays

mackworth committed:
Create new TestBed app to manipulate almost all parameters
Add coding to AverageLine
Move gradient properties to strong to properly retain
Fix compiler error about integers used in enums
Change alwaysDisplayPopupatIndex to use NSUInteger for index instead of float
Fix Frame warning in SimpleLineChart's launch Screen

mackworth committed:
Add colors, gradients, and alphas to TestBed
Fix many bugs and buglets

mackworth committed:
Fix assorted bugs in BEMSimpleLineGraph, especially null-data related
AverageLine:
•Color should be picked up from line if not set (default nil)
Typo in encoding macro

Line:
If null data and interpolation on, then extrapolate for beginning/ending nulls.
Bezier curve should be used even when only two points.
If interpolation off, then bezier line should be interrupted by gaps (but not top/bottom).
Avoid infinity result in midPoint calculation

SimpleLineGraphView:
Support restoration during startup.
NoDataLabel color should not default from Line (which defaults to white, same as background)
If null value, ensure corresponding Dot isn't left on chart
If label isn't used after initially being created, ensure it's removed from view
If neither X nor Y reference lines, set line's enableRefLines to NO (although default, might have been previously YES
If Xaxis background is defaulting to colorBottom, then also use alphaBottom to match; same for Yaxis and colorTop.
Avoid possible infinite loop if delegate gives a zero incrementIndex for x axis
Fix one-pixel gap between yaxis and graph
Remove spurious NSLogs

mackworth committed:
Allow use on iPhone and Split View

mackworth committed:
Fix bug with TouchLineInput color (doesn't change after initial setting)

mackworth committed:
Bug fixes:
Line clipsToBounds to avoid drawing outside the box (when extrapolating nulls or user set minValue too High)
Top/Bottom reference lines were outside box
When Bezier off and Interpolate nulls off, properly draws line segments now
Removed VMA NSLog from TestBed/MasterController

mackworth committed:
Supports bezier curves when line isn't being drawn
refactors bezier/linear processing into one method

mackworth committed:
Using cubic instead of quadratic Bezier curves
(Still has old code to allow comparison)

mackworth committed:
Simplifying code for controlPoints in cubicBezier
Using quad for background

mackworth committed:
Add variable X Axis support
Remove verticalReferenceHorizontalFringeNegation stuff
Fix background if points aren't on left/right edges
Adds new delegates: lineGraph:locationForPointAtIndex:  lineGraph:labelOnXAxisForLocation:,  numberOfXAxisLabelsOnLineGraph: maxXValueForLineGraph:, minXValueForLineGraph
Moves getting data  and calculations of points to new getData method

Adds new methods to TestBed
Adds missing icons to TestBed
Moves saving data from restoration to User defaults

mackworth committed:
Ensure all BEMnulls doesn't crash
Misc bug fixes in variable xAxis, esp minX/maxX
Add hints for minX/maxX in testbed
Add %nulls slider in testbed
Move from NSNotFound to -1 for empty markers in TestBed DetailVC

mackworth committed:
Remove prev  reference lines if axis turned off
No vert ref lines when value null (and not interpolated)
No Labels when dot is offscreen
In testBed, don't report dots when Null value
reset minX value when setting variableXaxis
in testBed restore font to fontname field

mackworth committed:
Remove prev  reference lines if axis turned off
No vert ref lines when value null (and not interpolated)
No Labels when dot is offscreen
In testBed, don't report dots when Null value
reset minX value when setting variableXaxis
in testBed restore font to fontname field

mackworth committed:
Remove prev  reference lines if axis turned off
No vert ref lines when value null (and not interpolated)
No Labels when dot is offscreen
In testBed, don't report dots when Null value
reset minX value when setting variableXaxis
in testBed restore font to fontname field

mackworth committed:
Implement zoom/pan for data
Double tap to toggle full scale and zoomed
Remove touchInput recognizers when disabled

mackworth committed:
Add user scaling (zoom/pan) to TestBed UI

mackworth committed:
Delegate callback upon scale change
Bug fix: turning scaling off didn't redraw properly
Moved zoom management to handleGesture

mackworth committed:
Further implementation of pan/zoom
Separate out two-finger pan from zoom recognizer for smoothness
Allow client setting scale/pan numbers (e.g. to sync two graphs)
Rename m--DisplayedValue to m--XDisplayedValue

mackworth committed:
Better handle overlapping labels with large numbers of points
Always display labels above dots
Check for all neighbors, not just two
Disappear dots outside of frame.

mackworth committed:
Add Area/Correlation to stats display

mackworth committed:
Major internal restructure
	Use Views to group dots and labels over BEMLine.
	Allows removal of kludgy yAxisOffset and xAxisOffset globals
	Set size of major views at beginning, so all calcs based on those views
	Changed all global mutable Arrays to NSArrays
	Used CGRectGetM... to simplify calculations
Fixes:
	Animation w/o displaying dots during left dots off previously
	Implemented numberOfXAxisLabelsOnLineGraph for indexed xAxis (in addition to previous location xAxis)
  	Added min 5 points between xAxisLabels
	Rewrite panZoom logic

mackworth committed:
Properties restructure:
Remove all global vars
Change all MutableArrays to Arrays
Remove ReferenceLine properties, hand directly to masterline instead
Rename AxisValues to Locations

mackworth committed:
Merge NewBezierAlgo into NoOffsets

mackworth committed:
Fix cubic merge glitch

mackworth committed:
Don't add off-screen points to line (fixes animation and faster)

mackworth committed:
Move to point array versus dual number arrays
Remove quadratic bezier algorithm

mackworth committed:
Implement xAxis at top Option
Add some missing property restorers?

mackworth committed:
Drop points off screen, but still include ones immediately offscreen for cubic bezier algorithm to work properly
After merge of cubic algorithm, found conflict with variable X axis work. Limited control points x-value to following point's x value.
Fixed startup problem for restoring scale and zoom
The usual random stuff from storyboard

mackworth committed:
Bug fixes:
   Some reference lines not appearing
   Sometimes highest xaxis label doesn't appear
   Nudge y labels onto page
   Highest y label doesn't appear when averagelabel is off
   When changing data sources, sometimes minXDisplayed gets out of sync
  Several properties not properly restored.

mackworth committed:
Convert back to NSInteger in API for Swift compatibility
Enable signed conversion checking and fix issues
Fix test for white/alpha to handle extended Grey space

mackworth committed:
Misc NSInteger type issues when compiling with strict and floats on 32 bit fr
Add EOL to some files
Fix for bezier edge of closed background (need to simplify)
New delegate to reformat yaxis labels
don't remove top level views when no data
Calc locations after sections laid out
Fix for right hand label out of bounds

mackworth committed:
Add ability to take screen shot bigger (or smaller) than screen
Move locationForPointAtIndex to datasource
Fix minor doc issues.
Fixed case where label is partially onscreen, but point is not.
Don't allow zooming infinitely small

mackworth committed:
Handle one datapoint in regular routines, not as special case
Simplify handling of zero points
Add NumberOfPoints field to TestBed
Fix bug when increasing numberofPoints by more than two
Add Activity Indicator to testBed to test didFinishDrawing

mackworth commented:
Supersedes PR #295, which also incorporates the changes mentioned above. I created a separate pull request in case the addition of the pan/zoom and variableX are determined to be inappropriate for this library.  Let me know what you think, especially on API additions, and/or if you have any problems with this build, and/or if there's any thing else you want me to implement, while I'm on a roll...

Boris-Em commented:
Thank you so much for your contributions @mackworth! As soon as we get your other PR (#295) merged in, I'll take a look at this one. If you plan on contributing more to the project, please keep on breaking your PRs as much as possible like you did.

mackworth committed:
PR #295 fixes
Spacing consistency with rest of project
topline fix
Move UIButton+switch and UITextField+Numbers from Master to separate files
Title of Master "Options"
Add Graph button if collapsed

mackworth committed:
Update .travis.yml

include fixes from other PR

mackworth committed:
Bugfixes:
Don't allow animation during redrawing for snapshot
Rounding issue on right edge for label/dots disappearing too soon
Better calculation for interpolating nulls on left edge of canvas during zoom

mackworth committed:
Merge remote-tracking branch 'origin/variableX' into variableX

mackworth committed:
Add adaptiveDataPoints property that lets calculations be based either on all datapoints (as it was and is the default) or just on currently displayed graph)
Change to labelOnXAxisForLocation to report which label on screen this is.
Align left-most and right-most labels left and right respecrtively.
Bug Fixes:
	Restore enableReferenceFrame functionality
	Use font's lineheight instead of pointsize to correctly calculate line height.
	Move zoomGestures to self rather than labelView to allow panning on XAxis
	Property remove zoomPanGesture (overlooked before)
	Avoid duplicate labels when delegate specified numberOfXAxisLabels is greater than actual number of labels
        Avoid some flickering due to rounding for dots on edge of screen.
	Avoid problem when switching back to indexed from location based axis
	In Master, Correctly load int and float defaults even when they are zero
	In Master, avoid some -1 issues.
	Some documentation typos

mackworth commented:
Added a few bug fixes. Also realized an ambiguity about calculations on graphDatapoints when you have pan/zoom. Does it mean calculate with the currently displayed points or all data points? I think both are useful, so I added a switch.

mackworth committed:
Better control over maxZoom
Double-tap restores zoom, but at double-tap location
Better behavior when sliding finger off screen.
Fix compiler error for nil font and NSInteger
Fix label tests

mackworth commented:
A couple bug fixes based on field testing of my app.
Noticed that the last set of commits broke the test, so undid one of my changes there.

Sam-Spencer commented:
Hi @mackworth, is there any way you can split out a separate pull request with **only** the pan / zoom feature - and nothing else?

There are a lot of changes here, and it is very difficult to accurately sift through everything. I do, however, have a few thoughts on what I have seen:

- Zoom and pan feature seems really nice in the brief tests I did.  
- Not sure how I feel about the cubic curve change from quadratic curves. Could be a matter of personal taste, but I don't like how this looks as much. While some regions of the graph with shallow slopes are displayed better, there's an odd kind of "*tilt*" to each large curve.  
- Some of these added and updated APIs would be much appreciated (in fact they would address *many* of the issues currently open on this project). However, my fear is that these changes to the public API may confuse or overburden `BEMSimpleLineGraphView` in its *current* state. Before continuing to add more features I think the project needs to be refactored. This would be quite the undertaking, but I believe that if done correctly it would make pull requests likes this much easier to merge, manage, and create in the first place. @Boris-Em what do you think? Maybe we can start a [refactoring project](https://github.com/Boris-Em/BEMSimpleLineGraph/projects) to discuss how to approach this?
